bundles/org.openhab.binding.folderwatcher/src/main/java/org/openhab/binding/folderwatcher/internal/handler/FtpFolderWatcherHandler.java

```diff
@@ -0,0 +1,254 @@
+            return;
+        }
+        if (config.pollInterval <= 0) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Polling interval can't be null or negative");
+        }
+
+        currentFtpListingFile = new File(OpenHAB.getUserDataFolder() + File.separator + "FolderWatcher" + File.separator
+                + thing.getUID().getAsString().replace(':', '_') + ".data");
+        try {
+            this.currentFtpListingFile = currentFtpListingFile;
+            previousFtpListing = WatcherCommon.initStorage(currentFtpListingFile, config.ftpAddress + config.ftpDir);
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+            logger.debug("Can't write file {}, error message {}", currentFtpListingFile, e.getMessage());
+            return;
+        }
+        this.initJob = scheduler.scheduleWithFixedDelay(this::connectionKeepAlive, 0, config.pollInterval,
+                TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void dispose() {
+        ScheduledFuture<?> executionJob = this.executionJob;
+        ScheduledFuture<?> initJob = this.initJob;
+        if (executionJob != null) {
+            executionJob.cancel(true);
+        }
+        if (initJob != null) {
+            initJob.cancel(true);
+        }
+        if (ftp.isConnected()) {
+            try {
+                ftp.logout();
+                ftp.disconnect();
+            } catch (IOException e) {
+                logger.debug("Error terminating FTP connection: ", e);
+            }
+        }
+    }
+
+    private List<String> listDirectory(FTPClient ftpClient, String parentDir, String currentDir, boolean recursive)
+            throws IOException {
+        Instant dateNow = Instant.now();
+        List<String> dirList = new ArrayList<>();
+        final String dirToList = parentDir + ((!currentDir.equals("")) ? ("/" + currentDir) : "");
+
+        for (FTPFile file : ftpClient.listFiles(dirToList)) {
+            String currentFileName = file.getName();
+            if (currentFileName.equals(".") || currentFileName.equals("..")) {
+                continue;
+            }
+            if (file.isDirectory()) {
+                if (recursive) {
+                    try {
+                        dirList.addAll(listDirectory(ftpClient, dirToList, currentFileName, recursive));
+                    } catch (IOException e) {
+                        logger.debug("Can't read FTP directory: {}", dirToList, e);
+                    }
+                }
+            } else {
+                long diff = ChronoUnit.HOURS.between(file.getTimestamp().toInstant(), dateNow);
+
+                if (diff < config.diffHours) {
+                    dirList.add("ftp:/" + ftpClient.getRemoteAddress().toString() + dirToList + "/" + currentFileName);
+                }
+            }
+        }
+        return dirList;
+    }
```

Comment:
It is better to have files appended to the same List instance so that you aren't creating a new instance for every directory you recurse. Passing a List instance as one of the recursive parameters would solve this. As an added bonus it seems that the logic using `listDirectory` is already trying to provide its own List instance to use so it would actually simplify the logic for the callers as well.

```suggestion
    private void listDirectory(FTPClient ftpClient, String dirPath, boolean recursive, List<String> dirFiles)  throws IOException {
        Instant dateNow = Instant.now();
        for (FTPFile file : ftpClient.listFiles(dirPath)) {
            String currentFileName = file.getName();
            if (currentFileName.equals(".") || currentFileName.equals("..")) {
                continue;
            }
            String filePath = dirPath + "/" + currentFileName;
            if (file.isDirectory()) {
                if (recursive) {
                    try {
                        listDirectory(ftpClient, filePath, recursive, dirFiles);
                    } catch (IOException e) {
                        logger.debug("Can't read FTP directory: {}", filePath, e);
                    }
                }
            } else {
                long diff = ChronoUnit.HOURS.between(file.getTimestamp().toInstant(), dateNow);
                if (diff < config.diffHours) {
                    dirList.add("ftp:/" + ftpClient.getRemoteAddress() + filePath);
                }
            }
        }
    }
```

Reply:
Done